### DISCUSSION TITLE
Grouped outputs support

### DISCUSSION THREAD

**Discussion by ein-shved** [YaLTeR/niri#866] [Ideas]
# Environment

I have setups: home office with single ultrawide screen and employer office with two equivalent 2k monitors placed side-by-side.

# Problem

I organised my work at home with stacked one-above-one workspaces with wide sliding window of output. But at work I have to use two independent stacks of workspaces, which is not very comfortable for me.

# Solution

I am thinking on creation the groups of outputs. Each group - is virtual output, with rectangle `((x0,y0),(x1,y1))`, where `(x0,y0)` - is top left corner of top left physical output, and `(x1,y1)` - is bottom right corner of bottom right output. 

Each workspaces now attaches to this virtual outputs (groups) instead of physical one. By default each physical output will be place to its own unique group.

### Output

I suggest to add new field to output section

```kdl

output "eDP-1" {
// ...
    group "AllMonitors"
// ...
}
```

This option will output `eDP-1` to group `AllMonitors`.

### Group

I suggest to add new section of type `group`, to allow user to configure their groups. There can be no such section for configured in `output` groups - in such case, the defaults will be applied.

```kdl

group "AllMonitors" {

    // How columns will synchronize their borders. Available values:
    //    * fit-current-physical - columns borders will be placed and calculated according to borders of current physical output
    //    * fit-maximum - columns borders will be placed and calculated according to full virtual output
    //    * fit-maximum - columns borders will be placed and calculated according to intersection of verticals of physical output
   column-fit "fit-current-physical"

    // How windows will synchronize their borders. Available values:
    //    * fit-current-physical - windows vertical borders will be placed and calculated according to borders of current physical output
    //    * fit-maximum - windows vertical borders will be placed and calculated according to full virtual output
    //    * fit-minimum - windows vertical borders will be placed and calculated according to intersection of verticals of physical output
   window-fit "fit-current-physical"
}

```

# Notes

This is will resolve #810

**Comment by ein-shved** [comment]
I am going to handle this by myself, but better to hear opinions of better informed people first =)

**Reply by YaLTeR** [reply]
This isn't the first time people ask for this, but I expect this to be very difficult.

You probably don't want a column to end up in the middle between two monitors; I'm honestly not sure how you'd handle this. The current layout code is built under the assumption that there are no constraints like this.

Even aside from that, lots of code in niri assumes that one workspace is on one output at a time. All kinds of Wayland events like scale, frame callbacks or dmabuf formats would need new handling.

Then there are design questions like what if the outputs don't exactly line up in size, or what if you try to group 3 outputs in an L shape.

**Reply by ein-shved** [reply]
> one workspace is on one output at a time

That exactly why I suggests the concept of `virtual output` with almost the same interface of physical. I am not talking that this will be easy but generally the concept will left the same.

> You probably don't want a column to end up in the middle between two monitors

I'am thinking of extra api of "sub-borders of physical outputs" in virtual outputs to handle this logic. I am familiar with the original code, but hope this will be possible.

> what if you try to group 3 outputs in an L shape.

This will be handled by configuration options of groups. There may not be ideal configuration model for each setup, but It could work well for generic layouts and may not needed to handle the exotics.
